### PR TITLE
[UI] fix frontend metric & config selection bugs + refactor

### PR DESCRIPTION
**Pull request by luca-dot-sh**
This PR refactors most of the frontend and fixes bugs.

Bugfixes:
- configurations being wrongly shared amogst PIDs
- metric selection only displaying global
- submit button being displayed if there are no changes
- Visualization Screen now only shows PIDs with configured programs and upon selecting a PID/App, only shows the configured options where there is actually data
- switch mode button is only shown if there is data being displayed.
-  visualization screen can now display errors

Refactorings:
- ConfigurationManager now tracks local (unconfirmed selection) and backend (actual active options) config and can be "synced" via submitting a config.
- DataStreamManager now uses SharedFlow to prevent multiple subscriptions
- Applications and Processes are provided by a combined component called RunningComponentProvider and Application Info is cached now
- VisualizationScreen now gets its state via a single data class
- VisualizationViewModel now does less copying but instead derives its information in a combined flow from the selected options, but still needs some improvements
- some renamings

Sorry for the size of this PR, but the bug fixes required some larger changes ...

Issues: #130 , #46

### CODE REVIEW

**Comment by fhilgers** [comment]
- [ ] TODO: fix visualization screen

**Review by fhilgers — CHANGES_REQUESTED** [review]
- [ ] Fix visualization screen

**Comment by luca-dot-sh** [comment]
The visualization should work now :)

**Comment by fhilgers** [comment]
Great, I'll check it out this afternoon!

**Comment by fhilgers** [comment]
I cannot confirm, either the event does not show at all, or it is only displayed very briefly:

[Screencast From 2024-12-03 15-22-47.webm](https://github.com/user-attachments/assets/f47d22dd-958f-4981-8327-24d98e0cbfb7)

**Comment by luca-dot-sh** [comment]
Weird, i will check it out again later this evening or tomorrow morning

**Comment by luca-dot-sh** [comment]
We were using a Flow instead of a StateFlow that wrongfully updated every second, leading to flickering. This was not present in the mocked version as its process list is constant.

File: frontend/app/src/main/java/de/amosproj3/ziofa/ui/visualization/VisualizationViewModel.kt [hunk 1]
```diff
@@ -55,29 +58,35 @@ class VisualizationViewModel(
...
+                    selectedTimeframe = activeTimeframe,
+                    componentOptions = configuredComponents.toUIOptions(),
+                    metricOptions =
+                        backendConfig.getMetricOptionsForPids(
+                            pids = activeComponent.getPIDsOrNull()
+                        ),
+                    timeframeOptions = if (activeMetric != null) DEFAULT_TIMEFRAME_OPTIONS else null,
+                )
+            }
+            .stateIn(viewModelScope, SharingStarted.Lazily, DEFAULT_SELECTION_DATA)
```
Comment (luca-dot-sh) [inline frontend/app/src/main/java/de/amosproj3/ziofa/ui/visualization/VisualizationViewModel.kt:89]: This is the important change.

**Review by fhilgers — APPROVED** [review]

**Pull request merged**